### PR TITLE
Address proxy behaviour yet again.

### PR DESCRIPTION
**Pull request by howbazaar**
This change updates develop to what was agreed at the last product sprint with regards to proxy behaviour. There was reasonable concern that we didn't want to break controllers that were working with the existing proxy behaviour.

This branch introduces a new set of model proxy values. These values are prefixed with "juju-" to indicate that they are not the normal proxy values.  A model can only have one set of proxy variables actually defined, either the legacy ones "http-proxy", "https-proxy", "ftp-proxy", or the new juju ones "juju-http-proxy", "juju-https-proxy", "juju-ftp-proxy". The config update validation code checks this.

The new juju proxy values are set in the agent environment, and added to the default http transport. The juju http transport uses a CIDR aware no-proxy.

In order to support the new proxy values, both the proxy updater and provisioner facades needed to be bumped, with appropriate changes in the api layer too.

Instance configuration also needed to take into account the change in behaviour for the initial cloud-init work.

## QA steps

The only real way to QA this will be to bootstrap within a proxy required environment and test behaviour. We have agreement with field that they'll test this.

## Documentation changes

Yes, we'll need to update docs on the proxy settings for 2.4.

### CODE REVIEW

File: apiserver/facades/agent/proxyupdater/proxyupdater.go [hunk 1]
```diff
@@ -16,6 +17,71 @@ import (
...
+	}
+	return NewAPIBase(
+		&stateShim{st: st, m: m},
+		ctx.Resources(),
+		ctx.Auth(),
+	)
+}
+
+// APIv1 provides the ProxyUpdater version 1 facade.
+type APIv1 struct {
```
Comment (wallyworld) [inline apiserver/facades/agent/proxyupdater/proxyupdater.go]: I think we mostly put these above the NewAPI() functions

File: apiserver/facades/agent/proxyupdater/proxyupdater.go [hunk 2]
```diff
@@ -16,6 +17,71 @@ import (
...
+	*APIv2
+}
+
+// APIv2 provides the ProxyUpdater version 2 facade.
+type APIv2 struct {
+	*APIBase
+}
+
+type APIBase struct {
```
Comment (wallyworld) [inline apiserver/facades/agent/proxyupdater/proxyupdater.go]: comment
or, does APIBase really need to be exported

File: apiserver/facades/agent/proxyupdater/proxyupdater.go [hunk 3]
```diff
@@ -25,25 +91,19 @@ type Backend interface {
...
-type ProxyUpdaterAPI struct {
-	backend    Backend
-	resources  facade.Resources
-	authorizer facade.Authorizer
-}
-
-// NewAPIWithBacking creates a new server-side API facade with the given Backing.
-func NewAPIWithBacking(st Backend, resources facade.Resources, authorizer facade.Authorizer) (*ProxyUpdaterAPI, error) {
+// NewAPIBase creates a new server-side API facade with the given Backing.
+func NewAPIBase(backend Backend, resources facade.Resources, authorizer facade.Authorizer) (*APIBase, error) {
```
Comment (wallyworld) [inline apiserver/facades/agent/proxyupdater/proxyupdater.go]: ditto here - this is internal now right?

File: apiserver/facades/agent/proxyupdater/proxyupdater.go [hunk 4]
```diff
@@ -124,15 +184,23 @@ func (api *ProxyUpdaterAPI) proxyConfig() params.ProxyConfigResult {
...
 
-	proxySettings := env.ProxySettings()
+	proxySettings := config.LegacyProxySettings()
 	proxySettings.AutoNoProxy = network.APIHostPortsToNoProxyString(apiHostPorts)
-	result.ProxySettings = proxyUtilsSettingsToProxySettingsParam(proxySettings)
-	result.APTProxySettings = proxyUtilsSettingsToProxySettingsParam(env.AptProxySettings())
+	result.LegacyProxySettings = toParams(proxySettings)
+
+	proxySettings = config.JujuProxySettings()
+	// Do we need the AutoNoProxy here?
```
Comment (wallyworld) [inline apiserver/facades/agent/proxyupdater/proxyupdater.go]: Can we make this a TODO to be addressed when the snap bits are added?

File: cloudconfig/userdatacfg_unix.go [hunk 5]
```diff
@@ -266,19 +265,25 @@ func (w *unixConfigure) ConfigureJuju() error {
 	w.conf.AddScripts(
 		// We look to see if the proxy line is there already as
 		// the manual provider may have had it already.
+		// We write this file out whether we are using the legacy proxy
+		// or the juju proxy to deal with runtime changes. The proxy updater worker
+		// only modifies /etc/juju-proxy.conf, so if changes are written to that file
+		// we need to make sure the profile.d file exists to reflect these changes.
+		// If the new juju proxies are used, the legacy proxies will be not set, and the
```
Comment (wallyworld) [inline cloudconfig/userdatacfg_unix.go]: will not be?
reads better?

File: environs/config/config.go [hunk 6]
```diff
@@ -100,6 +108,23 @@ const (
...
+	JujuHTTPSProxyKey = "juju-https-proxy"
+
+	// JujuFTPProxyKey stores the key for this setting.
+	JujuFTPProxyKey = "juju-ftp-proxy"
+
+	// JujuNoProxyKey stores the key for this setting.
+	JujuNoProxyKey = "juju-no-proxy"
+
 	// AptHTTPProxyKey stores the key for this setting.
 	AptHTTPProxyKey = "apt-http-proxy"
```
Comment (wallyworld) [inline environs/config/config.go]: To aid the reader, worth a comment why apt proxy is not legacy but is not given the JUJU_ prefix?

File: environs/config/config.go [hunk 7]
```diff
@@ -813,30 +850,81 @@ func (c *Config) ProxySettings() proxy.Settings {
...
+// HasJujuProxy returns true if there is any proxy set using the new juju-proxy keys.
+func (c *Config) HasJujuProxy() bool {
+	// We exclude the no proxy value as it has default value.
+	return c.JujuHTTPProxy() != "" ||
+		c.JujuHTTPSProxy() != "" ||
+		c.JujuFTPProxy() != ""
+}
+
+// JujuProxySettings returns all four proxy settings that have been set using the
+// juju- prefixed proxy settings. These values
```
Comment (wallyworld) [inline environs/config/config.go]: incomplete sentence

File: environs/config/config.go [hunk 8]
```diff
@@ -688,6 +718,11 @@ func Validate(cfg, old *Config) error {
 		}
 	}
 
+	// The user shouldn't specify both old and new proxy values.
+	if cfg.HasLegacyProxy() && cfg.HasJujuProxy() {
```
Comment (wallyworld) [inline environs/config/config.go]: I think we should also validate that juju no proxy, if set, contains valid cidr values

File: api/proxyupdater/proxyupdater.go [hunk 9]
```diff
@@ -74,12 +74,41 @@ func proxySettingsParamToProxySettings(cfg params.ProxyConfig) proxy.Settings {
 }
 
 // ProxyConfig returns the proxy settings for the current environment
-func (api *API) ProxyConfig() (proxySettings, APTProxySettings proxy.Settings, err error) {
+func (api *API) ProxyConfig() (legacyProxySettings, jujuProxySettings, aptProxySettings, snapProxySettings proxy.Settings, err error) {
```
[the remark this answers is not in the document]
Comment (howbazaar) [reply api/proxyupdater/proxyupdater.go]: I almost did this while writing the code, but was losing the will.

I'll relook at it.

File: apiserver/facades/agent/proxyupdater/proxyupdater.go (hunk 2 above)
Comment (howbazaar) [reply apiserver/facades/agent/proxyupdater/proxyupdater.go]: Not strictly, but it does mean we can use them without export_test wrappers.

**Comment by dshcherb** [comment]
Quick question: is the no-proxy code aware of wildcards for hostnames similar to
https://go-review.googlesource.com/c/go/+/75730/3/src/net/http/no_proxy.go#159 ?

**Comment by howbazaar** [comment]
@dshcherb I don't know. The CIDR aware proxy is in the utils/proxy package of this codebase.

**Comment by howbazaar** [comment]
$$merge$$

**Pull request merged**